octodns/record/chunked.py

```diff
@@ -34,6 +34,7 @@ def rr_values(self):
 
 class _ChunkedValue(str):
     _unescaped_semicolon_re = re.compile(r'\w;')
+    _chunk_sep_re = re.compile(r'"\s+"')
```

Comment:
Strictly speaking, the spec would also allow "mixed case chunks" such as `"foo bar baz" blablablablabla "yabba dabba dooh"`. Do we want to handle this as well or wait until we encounter this in the wild?

Reply:
In any case I would change the implementation to properly tokenize `character-string` as per spec:
- Treat the leading and trailing " to be part of each chunk, not `"\s+"` to be a separator.
- Check for wellformedness (unescaped " pair matching).
- Unescape embedded `\"` as per spec.

Due to the definition of `character-string` I guess that a walk/state-based approach would work better than a regex based approach:

1. set `target_value` to empty
2. walk whitespace (i.e. strip leading whitespace)
3. set `chunk_start_pos` to `pos`
4. set `quoted_flag` if `v[pos] == '"'` else reset `quoted_flag`
5. walk to next `"` or `\s` depending on quoted flag
6. if `quoted_flag and '\' == v[pos-1]`: append to  `target_value` from `chunk_start_pos` to `pos - 2`, repeat from 5. (this means: keep walking the current quoted chunk)
7. append to  `target_value` from `chunk_start_pos` to `pos - 1 (this means: finish any current chunk)
8. repeat from 2.
9. throw error if `quoted_flag and '"' != v[-1]`

If you want to stick with the simple regex approach, I would tend to use `"\s*"` to be the separator rather than `\s+`.

Reply:
> Strictly speaking, the spec would also allow "mixed case chunks" such as `"foo bar baz" blablablablabla "yabba dabba dooh"`. Do we want to handle this as well or wait until we encounter this in the wild?

Interesting. 

> In any case I would change the implementation to properly tokenize `character-string` as per spec:
> ...

Will explore this, and probably check and see what other DNS libs do for it as well.

Reply:
> Strictly speaking, the spec would also allow "mixed case chunks" such as `"foo bar baz" blablablablabla "yabba dabba dooh"`. Do we want to handle this as well or wait until we encounter this in the wild?

Double checking my current read here. Unquoted values cannot contain spaces? The old code definitely allowed that, but tbh no clue if it actually ever saw it in the real world. Parser-wise I'm not really sure how we'd support mixing quoted and unquoted if the unquoted bits can have spaces.

Reply:
Figured the best/easiest thing to do would be to turn to octodns-bind and zonefile provider to see how it behaves since it's using dnspython for reading the zonefile:

Zonefile
```zonefile
one IN  TXT no quotes in this one
two IN  TXT "this one was quoted"
three IN  TXT "this one has both, this quoted bit" and this unquoted section
four IN  TXT "multiple quoted" "sections in this one"
five IN  TXT "same with multiple spaces"     "between the quotes"
six IN  TXT   "leading spaces before" "and trailing spaces after"    
seven IN  TXT   "first value"
seven IN  TXT   "second value"
```

Data loaded from that

```python
[Rr<exxampled.com., NS, 86400, dns1.exxampled.com.,
 Rr<exxampled.com., NS, 86400, dns2.exxampled.com.,
 Rr<one.exxampled.com., TXT, 86400, "no" "quotes" "in" "this" "one",
 Rr<two.exxampled.com., TXT, 86400, "this one was quoted",
 Rr<three.exxampled.com., TXT, 86400, "this one has both, this quoted bit" "and" "this" "unquoted" "section",
 Rr<four.exxampled.com., TXT, 86400, "multiple quoted" "sections in this one",
 Rr<five.exxampled.com., TXT, 86400, "same with multiple spaces" "between the quotes",
 Rr<six.exxampled.com., TXT, 86400, "leading spaces before" "and trailing spaces after",
 Rr<seven.exxampled.com., TXT, 86400, "first value",
 Rr<seven.exxampled.com., TXT, 86400, "second value"]
```

And what octoDNS currently makes of that:

```console
********************************************************************************
* exxampled.com.
********************************************************************************
* config (YamlProvider)
*   Create Zone<exxampled.com.>
*   Create <NsRecord NS 86400, exxampled.com., ['dns1.exxampled.com.', 'dns2.exxampled.com.']> ()
*   Create <TxtRecord TXT 86400, five.exxampled.com., ['same with multiple spacesbetween the quotes']> ()
*   Create <TxtRecord TXT 86400, four.exxampled.com., ['multiple quotedsections in this one']> ()
*   Create <TxtRecord TXT 86400, one.exxampled.com., ['noquotesinthisone']> ()
*   Create <TxtRecord TXT 86400, seven.exxampled.com., ['first value', 'second value']> ()
*   Create <TxtRecord TXT 86400, six.exxampled.com., ['leading spaces beforeand trailing spaces after']> ()
*   Create <TxtRecord TXT 86400, three.exxampled.com., ['this one has both, this quoted bitandthisunquotedsection']> ()
*   Create <TxtRecord TXT 86400, two.exxampled.com., ['this one was quoted']> ()
*   Summary: Creates=8, Updates=0, Deletes=0, Existing Records=0
********************************************************************************
```

That pretty well lays out the scope of possible behavior. 

It is clear that octodns-bind wouldn't round trip unquoted items. I think results would semantically be the same, but the actual text written out file will all be quoted.:

```zonefile
five     86400 IN TXT      "same with multiple spacesbetween the quotes"
four     86400 IN TXT      "multiple quotedsections in this one"
one      86400 IN TXT      "nospacesinthisone"
seven    86400 IN TXT      "first value"
         86400 IN TXT      "second value"
six      86400 IN TXT      "leading spaces beforeand trailing spaces after"
three    86400 IN TXT      "this one has both, this quoted bitandthisunquotedsection"
two      86400 IN TXT      "this one was quoted"
```

I don't see a way this could be avoided since octoDNS internally represents a value as an single arbitrary length string which I think is FAR preferable to avoid every user having to deal will all the complexities that we're no looking at.

So my plan for a path forward is to explore using dnspython's code for parsing the data. I remember looking into it back when I first did all the rdata stuff, but didn't end up using it and I don't recall exactly why. If that doesn't end up making sense I'll look at a parser setup that will match the behavior.

Reply:
So I have a setup that's using dnspython to do the handling of the strings. It's not super straightforward since dnspython internally holds things as a list of `bytes`, but that can be decoded into what octoDNS needs. The other "extra" bit here is that dnspython does not escape `;`, which is one of the biggest regrets I have with octoDNS. I so wish I didn't decide that they needed to be escaped in TXT values. I just put it on my TODO list to explore whether we can get rid of that as part of a major release, but ...

A side effect of all this digging work is that a bunch of the existing TXT rdata test cases, all the ones w/o quotes, fail with the changes as there was an incorrect assumption that whitespace should be preserved. I don't think any rdata commonly works with unquoted TXT values and looking at the code/uses I think that change is fine, but just wanted to call it out/record it here.

Reply:
Scratch that, dnspython route isn't viable as there are validation checks preventing it from processing tokens > 255 chars. I can't feed it pieces of the value since the parsing looking for quotes would span the chunks.